cocotb/triggers.py

```diff
@@ -194,7 +194,19 @@ def __init__(self, time_ps, units=None):
 
         See Also:
             :func:`~cocotb.utils.get_sim_steps`
+
+        Raises:
+            TriggerException: If a negative value is passed for Timer setup.
+
+        .. versionadded:: 1.5
+            Raise an exception when Timer uses a negative value as it is undefined behavior.
+            Warn for 0 as this will cause erratic behavior in some simulators as well.
         """
+        if time_ps <= 0:
+            if time_ps == 0:
+                self.log.warning("Timer setup with value 0, which might exhibit undefined behavior in some simulators")
```

Comment:
This will fail because `self.log` doesn't exist yet.

I think I'd lean towards using `warnings.warn("...", stacklevel=2)` here instead.

Reply:
It did work here in Travis CI:
![grafik](https://user-images.githubusercontent.com/25175069/95452175-28767500-0969-11eb-9256-8308640d86d6.png)

Reply:
Ah, you're right - I changed how this worked

The testability argument still applies though. #1773 would make my suggestion appear as a normal log line too.

Reply:
I'm not quite sure I get your point, don't we have a test right now with what I have pasted above?

Or do you have something else/additional in mind?

(feel free to make the edit or give the exact line you would like to see, I have no real preference)

Reply:
It's not a proper test if it can't fail :). You can use `warnings.catch_warnings` or similar to record what warnings happen, and then `assert` the expected one is present.

Reply:
```suggestion
               warnings.warn(
                   message="Timer setup with value 0, which might exhibit undefined behavior in some simulators", 
                   category=RuntimeWarning,
                   stacklevel=2)
```

You will need to import `warnings` at the top.

Reply:
Mark, you will see the codecov bot posting coverage results. It shows certain paths not being tested. If you add a new feature, you need to add a test to cover your addition, at a minimum. See the `tests/test_cases/test_cocotb` directory. There are a couple examples of tests that catch warnings in `test_deprecated`. I suggest you use the `assert_raises` context manager defined there in your test as well. Those tests should go in `test_timing_triggers.py` in that directory.

Reply:
I'm sorry this isn't better spelt out yet, we are working on it (#2090).

Reply:
Yep, so for the warning we should be set.
I decided to do it directly in this helper function, so the check will be carried out whenever this is used.

Reply:
Marking as unresolved because you need @ktbarrett's suggested code